zebra-rpc/src/methods.rs

```diff
@@ -518,6 +537,88 @@ where
         }
         .boxed()
     }
+
+    fn z_get_treestate(&self, hash_or_height: String) -> BoxFuture<Result<GetTreestate>> {
+        let mut state = self.state.clone();
+
+        async move {
+            let hash_or_height = hash_or_height
+                .parse()
+                .map_err(|error: SerializationError| Error {
+                    code: ErrorCode::ServerError(0),
+                    message: error.to_string(),
+                    data: None,
+                })?;
+
+            let block_request = zebra_state::ReadRequest::Block(hash_or_height);
+            let block_response = state
+                .ready()
+                .and_then(|service| service.call(block_request))
+                .await
+                .map_err(|error| Error {
+                    code: ErrorCode::ServerError(0),
+                    message: error.to_string(),
+                    data: None,
+                })?;
+
+            let time = match block_response {
+                zebra_state::ReadResponse::Block(Some(block)) => block.header.time,
+                zebra_state::ReadResponse::Block(None) => {
+                    return Err(Error {
+                        code: ErrorCode::ServerError(0),
+                        message: "could not get the time of the requested block".to_string(),
+                        data: None,
+                    })
+                }
+                _ => unreachable!("unmatched response to a block request"),
+            };
+
+            let sapling_request = zebra_state::ReadRequest::SaplingTree(hash_or_height);
+            let sapling_response = state
+                .ready()
+                .and_then(|service| service.call(sapling_request))
+                .await
+                .map_err(|error| Error {
+                    code: ErrorCode::ServerError(0),
+                    message: error.to_string(),
+                    data: None,
+                })?;
+
+            let sapling_tree = match sapling_response {
+                zebra_state::ReadResponse::SaplingTree(Some(tree)) => {
+                    NoteCommitmentTree::Sapling(Some((*tree).clone()))
+                }
+                zebra_state::ReadResponse::SaplingTree(None) => NoteCommitmentTree::Sapling(None),
+                _ => unreachable!("unmatched response to a sapling tree request"),
+            };
+
+            let orchard_request = zebra_state::ReadRequest::SaplingTree(hash_or_height);
```

Comment:
I'm guessing this request type will get fixed later?

Reply:
No, this was a mistake! :D I fixed it.